example-app/src/main/java/com/adyen/checkout/example/ui/settings/viewmodel/SettingsViewModel.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2024 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by josephj on 3/9/2024.
+ */
+
+package com.adyen.checkout.example.ui.settings.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.adyen.checkout.example.ui.settings.model.EditSettingsData
+import com.adyen.checkout.example.ui.settings.model.SettingsIdentifier
+import com.adyen.checkout.example.ui.settings.model.SettingsItem
+import com.adyen.checkout.example.ui.settings.model.SettingsUIState
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+internal class SettingsViewModel @Inject constructor(
+    private val settingsUIMapper: SettingsUIMapper,
+    private val settingsEditor: SettingsEditor,
+) : ViewModel() {
+
+    private val _uiState = MutableStateFlow(SettingsUIState(emptyList()))
+    val uiState: StateFlow<SettingsUIState> = _uiState.asStateFlow()
+
+    init {
+        fetchSettings()
+    }
+
+    fun onItemClicked(item: SettingsItem) {
+        val editSettingsData = settingsUIMapper.getEditSettingsData(item)
+        updateUIState {
+            it.copy(
+                settingToEdit = editSettingsData,
+            )
+        }
+    }
+
+    fun onEditSettingConsumed() {
+        updateUIState {
+            it.copy(
+                settingToEdit = null,
+            )
+        }
+    }
+
+    fun onTextSettingChanged(identifier: SettingsIdentifier, newValue: String) {
+        settingsEditor.editSetting(identifier, newValue)
+        fetchSettings()
+    }
+
+    fun onListSettingChanged(identifier: SettingsIdentifier, selectedItem: EditSettingsData.SingleSelectList.Item) {
+        settingsEditor.editSetting(identifier, selectedItem)
+        fetchSettings()
+    }
+
+    fun onSwitchSettingChanged(identifier: SettingsIdentifier, newValue: Boolean) {
+        settingsEditor.editSetting(identifier, newValue)
+        fetchSettings()
+    }
```

Comment:
What do you think of only leaving one `onSettingChanged()` function here, which would accept an `identifier` and `value` (with type `EditSettingsData`), which could be passed to the `settingsEditor`, which in turn would distinguish the differences between setting items.

This way if we add or remove new type of settings, we do not need to modify the `SettingsViewModel`.

Reply:
Check out the last 2 commits and let me know what do you think

Reply:
Looks good! 👍